File: Services/LocalizationManager.vb
Imports System.Collections.ObjectModel
Imports System.Globalization
Imports System.Threading
Imports System.Windows
Imports System.Windows.Threading
Imports WpfApp1.Models

Namespace Services
    Public NotInheritable Class LocalizationManager
        Private Const DefaultCultureName As String = "en-US"
        Private Const ChineseCultureName As String = "zh-CN"

        Private Shared ReadOnly _instance As New LocalizationManager()

        Private ReadOnly _availableLanguages As ReadOnlyCollection(Of LanguageOption)
        Private _activeDictionary As ResourceDictionary
        Private _currentLanguage As LanguageOption
        Private _initialized As Boolean

        Private Sub New()
            _availableLanguages =
                New ReadOnlyCollection(Of LanguageOption)(
                    New List(Of LanguageOption) From {
                        New LanguageOption("zh-CN", "简体中文", "Chinese"),
                        New LanguageOption("en-US", "English", "English")
                    })
        End Sub

        Public Shared ReadOnly Property Instance As LocalizationManager
            Get
                Return _instance
            End Get
        End Property

        Public Event LanguageChanged As EventHandler

        Public ReadOnly Property AvailableLanguages As ReadOnlyCollection(Of LanguageOption)
            Get
                Return _availableLanguages
            End Get
        End Property

        Public ReadOnly Property CurrentLanguage As LanguageOption
            Get
                Return _currentLanguage
            End Get
        End Property

        Public ReadOnly Property CurrentCulture As CultureInfo
            Get
                If _currentLanguage Is Nothing Then
                    Return CultureInfo.GetCultureInfo(DefaultCultureName)
                End If

                Return CultureInfo.GetCultureInfo(_currentLanguage.CultureName)
            End Get
        End Property

        Public Sub Initialize()
            If _initialized Then
                Return
            End If

            EnsureWpfResourceContext()
            _initialized = True
            SetLanguage(ResolveCultureName(CultureInfo.CurrentUICulture.Name), False)
        End Sub

        Public Function SetLanguage(cultureName As String, Optional raiseChanged As Boolean = True) As Boolean
            EnsureWpfResourceContext()

            Dim resolvedLanguage = FindLanguage(cultureName)
            If resolvedLanguage Is Nothing Then
                resolvedLanguage = FindLanguage(DefaultCultureName)
            End If

            Dim requiresReload =
                _activeDictionary Is Nothing OrElse
                _currentLanguage Is Nothing OrElse
                Not String.Equals(_currentLanguage.CultureName, resolvedLanguage.CultureName, StringComparison.OrdinalIgnoreCase)

            If requiresReload Then
                Dim dictionary = LoadLanguageDictionary(resolvedLanguage.CultureName)
                ReplaceLanguageDictionary(dictionary)
                _currentLanguage = resolvedLanguage
            End If

            Dim targetCulture = CultureInfo.GetCultureInfo(_currentLanguage.CultureName)
            Thread.CurrentThread.CurrentCulture = targetCulture
            Thread.CurrentThread.CurrentUICulture = targetCulture

            If raiseChanged Then
                RaiseEvent LanguageChanged(Me, EventArgs.Empty)
            End If

            Return requiresReload
        End Function

        Public Function GetString(key As String, ParamArray args() As Object) As String
            If String.IsNullOrWhiteSpace(key) Then
                Return String.Empty
            End If

            If Not _initialized Then
                Initialize()
            End If

            Dim value = ResolveResourceValue(key)
            If String.IsNullOrWhiteSpace(value) Then
                Return key
            End If

            If args Is Nothing OrElse args.Length = 0 Then
                Return value
            End If

            Return String.Format(CurrentCulture, value, args)
        End Function

        Private Function ResolveResourceValue(key As String) As String
            If _activeDictionary IsNot Nothing AndAlso _activeDictionary.Contains(key) Then
                Return TryCast(_activeDictionary(key), String)
            End If

            If Application.Current IsNot Nothing AndAlso Application.Current.Resources.Contains(key) Then
                Return TryCast(Application.Current.Resources(key), String)
            End If

            Return Nothing
        End Function

        Private Function LoadLanguageDictionary(cultureName As String) As ResourceDictionary
            Dim languageDictionary As New ResourceDictionary()
            languageDictionary.MergedDictionaries.Add(New ResourceDictionary With {
                .Source = BuildComponentUri(String.Format(CultureInfo.InvariantCulture,
                                                          "Resources/Typography/Typography.{0}.xaml",
                                                          cultureName))
            })
            languageDictionary.MergedDictionaries.Add(New ResourceDictionary With {
                .Source = BuildComponentUri(String.Format(CultureInfo.InvariantCulture,
                                                          "Resources/Localization/Strings.{0}.xaml",
                                                          cultureName))
            })
            Return languageDictionary
        End Function

        Private Sub ReplaceLanguageDictionary(dictionary As ResourceDictionary)
            If dictionary Is Nothing OrElse Application.Current Is Nothing Then
                _activeDictionary = dictionary
                Return
            End If

            If _activeDictionary IsNot Nothing Then
                Application.Current.Resources.MergedDictionaries.Remove(_activeDictionary)
            End If

            _activeDictionary = dictionary
            Application.Current.Resources.MergedDictionaries.Add(_activeDictionary)
        End Sub

        Private Function FindLanguage(cultureName As String) As LanguageOption
            If String.IsNullOrWhiteSpace(cultureName) Then
                Return Nothing
            End If

            Dim exact = _availableLanguages.FirstOrDefault(Function(item) String.Equals(item.CultureName, cultureName, StringComparison.OrdinalIgnoreCase))
            If exact IsNot Nothing Then
                Return exact
            End If

            Dim languagePrefix = cultureName.Split("-"c)(0)
            Return _availableLanguages.FirstOrDefault(Function(item) item.CultureName.StartsWith(languagePrefix, StringComparison.OrdinalIgnoreCase))
        End Function

        Private Shared Function ResolveCultureName(cultureName As String) As String
            If IsChineseSystemCulture(cultureName) Then
                Return ChineseCultureName
            End If

            Return DefaultCultureName
        End Function

        Private Shared Function IsChineseSystemCulture(cultureName As String) As Boolean
            If String.IsNullOrWhiteSpace(cultureName) Then
                Return False
            End If

            Dim normalizedCultureName = cultureName.Trim()

            If normalizedCultureName.Equals("zh", StringComparison.OrdinalIgnoreCase) Then
                Return True
            End If

            If normalizedCultureName.StartsWith("zh-Hans", StringComparison.OrdinalIgnoreCase) OrElse
               normalizedCultureName.StartsWith("zh-Hant", StringComparison.OrdinalIgnoreCase) OrElse
               normalizedCultureName.Equals("zh-CHS", StringComparison.OrdinalIgnoreCase) OrElse
               normalizedCultureName.Equals("zh-CHT", StringComparison.OrdinalIgnoreCase) Then
                Return True
            End If

            Try
                Dim culture = CultureInfo.GetCultureInfo(normalizedCultureName)
                If Not String.Equals(culture.TwoLetterISOLanguageName, "zh", StringComparison.OrdinalIgnoreCase) Then
                    Return False
                End If

                Dim regionName = culture.Name.Split("-"c).LastOrDefault()
                Return String.Equals(regionName, "CN", StringComparison.OrdinalIgnoreCase) OrElse
                       String.Equals(regionName, "SG", StringComparison.OrdinalIgnoreCase) OrElse
                       String.Equals(regionName, "TW", StringComparison.OrdinalIgnoreCase) OrElse
                       String.Equals(regionName, "HK", StringComparison.OrdinalIgnoreCase) OrElse
                       String.Equals(regionName, "MO", StringComparison.OrdinalIgnoreCase)
            Catch ex As CultureNotFoundException
                Return False
            End Try
        End Function

        Private Shared Function BuildComponentUri(relativePath As String) As Uri
            Dim assemblyName = GetType(LocalizationManager).Assembly.GetName().Name
            Dim source = String.Format(CultureInfo.InvariantCulture,
                                       "pack://application:,,,/{0};component/{1}",
                                       assemblyName,
                                       relativePath.Replace("\"c, "/"c))
            Return New Uri(source, UriKind.Absolute)
        End Function

        Private Shared Sub EnsureWpfResourceContext()
            If Application.Current IsNot Nothing Then
                Dim applicationDispatcher As Dispatcher = Application.Current.Dispatcher
                Return
            End If

            Dim threadDispatcher As Dispatcher = Dispatcher.CurrentDispatcher
        End Sub
    End Class
End Namespace

File: src/ClickSyncMouseTester/Services/LocalizationManager.cs
using ClickSyncMouseTester.Models;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Globalization;
using System.Linq;
using System.Threading;
using System.Windows;
using System.Windows.Threading;

namespace ClickSyncMouseTester.Services;

public sealed class LocalizationManager
{
    private const string DefaultCultureName = "en-US";
    private const string ChineseCultureName = "zh-CN";

    private static readonly LocalizationManager _instance = new LocalizationManager();

    private readonly ReadOnlyCollection<LanguageOption> _availableLanguages;
    private ResourceDictionary _activeDictionary;
    private LanguageOption _currentLanguage;
    private bool _initialized;

    public static LocalizationManager Instance => _instance;

    public ReadOnlyCollection<LanguageOption> AvailableLanguages => _availableLanguages;

    public LanguageOption CurrentLanguage => _currentLanguage;

    public CultureInfo CurrentCulture => CultureInfo.GetCultureInfo(_currentLanguage?.CultureName ?? DefaultCultureName);

    public event EventHandler LanguageChanged;

    private LocalizationManager()
    {
        _availableLanguages = new ReadOnlyCollection<LanguageOption>(new List<LanguageOption>
        {
            new LanguageOption(ChineseCultureName, "简体中文", "Chinese"),
            new LanguageOption(DefaultCultureName, "English", "English")
        });
    }

    public void Initialize()
    {
        if (_initialized)
        {
            return;
        }

        EnsureWpfResourceContext();
        _initialized = true;
        SetLanguage(ResolveCultureName(CultureInfo.CurrentUICulture.Name), raiseChanged: false);
    }

    public bool SetLanguage(string cultureName, bool raiseChanged = true)
    {
        EnsureWpfResourceContext();

        LanguageOption targetLanguage = FindLanguage(cultureName) ?? FindLanguage(DefaultCultureName);
        bool languageChanged = _activeDictionary == null
            || _currentLanguage == null
            || !string.Equals(_currentLanguage.CultureName, targetLanguage.CultureName, StringComparison.OrdinalIgnoreCase);

        if (languageChanged)
        {
            ResourceDictionary languageDictionary = LoadLanguageDictionary(targetLanguage.CultureName);
            ReplaceLanguageDictionary(languageDictionary);
            _currentLanguage = targetLanguage;
        }

        CultureInfo targetCulture = CultureInfo.GetCultureInfo(_currentLanguage.CultureName);
        Thread.CurrentThread.CurrentCulture = targetCulture;
        Thread.CurrentThread.CurrentUICulture = targetCulture;

        if (raiseChanged)
        {
            LanguageChanged?.Invoke(this, EventArgs.Empty);
        }

        return languageChanged;
    }

    public string GetString(string key, params object[] args)
    {
        if (string.IsNullOrWhiteSpace(key))
        {
            return string.Empty;
        }

        if (!_initialized)
        {
            Initialize();
        }

        string localizedText = ResolveResourceValue(key);
        if (string.IsNullOrWhiteSpace(localizedText))
        {
            return key;
        }

        return args == null || args.Length == 0
            ? localizedText
            : string.Format(CurrentCulture, localizedText, args);
    }

    private string ResolveResourceValue(string key)
    {
        if (_activeDictionary != null && _activeDictionary.Contains(key))
        {
            return _activeDictionary[key] as string;
        }

        if (System.Windows.Application.Current != null && System.Windows.Application.Current.Resources.Contains(key))
        {
            return System.Windows.Application.Current.Resources[key] as string;
        }

        return null;
    }

    private ResourceDictionary LoadLanguageDictionary(string cultureName)
    {
        string resourceCultureName = cultureName;

        return new ResourceDictionary
        {
            MergedDictionaries =
            {
                new ResourceDictionary
                {
                    Source = BuildResourceUri(string.Format(CultureInfo.InvariantCulture, "/Resources/Typography/Typography.{0}.xaml", resourceCultureName))
                },
                new ResourceDictionary
                {
                    Source = BuildResourceUri(string.Format(CultureInfo.InvariantCulture, "/Resources/Localization/Strings.{0}.xaml", resourceCultureName))
                }
            }
        };
    }

    private void ReplaceLanguageDictionary(ResourceDictionary languageDictionary)
    {
        if (languageDictionary == null || System.Windows.Application.Current == null)
        {
            _activeDictionary = languageDictionary;
            return;
        }

        if (_activeDictionary != null)
        {
            System.Windows.Application.Current.Resources.MergedDictionaries.Remove(_activeDictionary);
        }

        _activeDictionary = languageDictionary;
        System.Windows.Application.Current.Resources.MergedDictionaries.Add(_activeDictionary);
    }

    private LanguageOption FindLanguage(string cultureName)
    {
        if (string.IsNullOrWhiteSpace(cultureName))
        {
            return null;
        }

        LanguageOption exactMatch = _availableLanguages.FirstOrDefault(language => string.Equals(language.CultureName, cultureName, StringComparison.OrdinalIgnoreCase));
        if (exactMatch != null)
        {
            return exactMatch;
        }

        string languagePrefix = cultureName.Split('-')[0];
        return _availableLanguages.FirstOrDefault(language => language.CultureName.StartsWith(languagePrefix, StringComparison.OrdinalIgnoreCase));
    }

    private static string ResolveCultureName(string cultureName)
    {
        return IsChineseSystemCulture(cultureName) ? ChineseCultureName : DefaultCultureName;
    }

    private static bool IsChineseSystemCulture(string cultureName)
    {
        if (string.IsNullOrWhiteSpace(cultureName))
        {
            return false;
        }

        string normalizedCultureName = cultureName.Trim();
        if (normalizedCultureName.Equals("zh", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        if (normalizedCultureName.StartsWith("zh-Hans", StringComparison.OrdinalIgnoreCase)
            || normalizedCultureName.StartsWith("zh-Hant", StringComparison.OrdinalIgnoreCase)
            || normalizedCultureName.Equals("zh-CHS", StringComparison.OrdinalIgnoreCase)
            || normalizedCultureName.Equals("zh-CHT", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        try
        {
            CultureInfo culture = CultureInfo.GetCultureInfo(normalizedCultureName);
            if (!string.Equals(culture.TwoLetterISOLanguageName, "zh", StringComparison.OrdinalIgnoreCase))
            {
                return false;
            }

            string regionName = culture.Name.Split('-').LastOrDefault();
            return string.Equals(regionName, "CN", StringComparison.OrdinalIgnoreCase)
                || string.Equals(regionName, "SG", StringComparison.OrdinalIgnoreCase)
                || string.Equals(regionName, "TW", StringComparison.OrdinalIgnoreCase)
                || string.Equals(regionName, "HK", StringComparison.OrdinalIgnoreCase)
                || string.Equals(regionName, "MO", StringComparison.OrdinalIgnoreCase);
        }
        catch (CultureNotFoundException)
        {
            return false;
        }
    }

    private static Uri BuildResourceUri(string resourcePath)
    {
        string normalizedPath = resourcePath.Replace('\\', '/');
        return new Uri(normalizedPath, UriKind.Relative);
    }

    private static void EnsureWpfResourceContext()
    {
        if (System.Windows.Application.Current != null)
        {
            _ = System.Windows.Application.Current.Dispatcher;
            return;
        }

        _ = Dispatcher.CurrentDispatcher;
    }
}
